File: DevSupportAndroid/src/main/java/com/sup/dev/android/libs/api_simple/ApiRequestsSupporter.java
package com.sup.dev.android.libs.api_simple;

import android.support.annotation.StringRes;

import com.sup.dev.android.R;
import com.sup.dev.android.app.SupAndroid;
import com.sup.dev.android.libs.screens.NavigationAction;
import com.sup.dev.android.libs.screens.Navigator;
import com.sup.dev.android.libs.screens.Screen;
import com.sup.dev.android.tools.ToolsResources;
import com.sup.dev.android.tools.ToolsToast;
import com.sup.dev.android.tools.ToolsView;
import com.sup.dev.android.views.screens.SAlert;
import com.sup.dev.android.views.screens.SInterstitialProgress;
import com.sup.dev.android.views.widgets.Widget;
import com.sup.dev.android.views.widgets.WidgetAlert;
import com.sup.dev.android.views.widgets.WidgetProgressTransparent;
import com.sup.dev.android.views.widgets.WidgetProgressWithTitle;
import com.sup.dev.java.classes.callbacks.simple.Callback1;
import com.sup.dev.java.classes.callbacks.simple.Callback2;
import com.sup.dev.java.classes.providers.Provider1;
import com.sup.dev.java.libs.api_simple.client.ApiClient;
import com.sup.dev.java.libs.api_simple.client.Request;
import com.sup.dev.java.tools.ToolsDate;
import com.sup.dev.java.tools.ToolsThreads;

public class ApiRequestsSupporter {
    
    private static ApiClient api;
    
    public static void init(ApiClient api){
        ApiRequestsSupporter.api = api;
    }


    public static <K extends Request.Response> Request<K> executeInterstitial(NavigationAction action, Request<K> request, Provider1<K, Screen> onComplete) {

        SInterstitialProgress sInterstitialProgress = new SInterstitialProgress();
        Navigator.action(action, sInterstitialProgress);

        long startTime = System.currentTimeMillis();
        request.onComplete(
                r -> {
                    if (Navigator.getCurrent() == sInterstitialProgress) {
                        long timeLeft = System.currentTimeMillis() - startTime;
                        if (timeLeft >= 300) Navigator.replace(onComplete.provide(r));
                        else ToolsThreads.main(300 - timeLeft, () -> {
                            if (Navigator.getCurrent() == sInterstitialProgress) Navigator.replace(onComplete.provide(r));
                        });
                    }

                })
                .onNetworkError(() -> {
                    if (Navigator.getCurrent() == sInterstitialProgress) {
                        SAlert.showNetwork(Navigator.REPLACE, () -> {
                            Navigator.replace(sInterstitialProgress);
                            request.send(api);
                        });
                    }
                })
                .onApiError(ApiClient.ERROR_GONE, e -> {
                    if (Navigator.getCurrent() == sInterstitialProgress) SAlert.showGone(Navigator.REPLACE);
                })
                .onApiError(ApiClient.ERROR_ACCOUNT_IS_BANED, ex -> ToolsToast.show(String.format(ToolsResources.getString(SupAndroid.TEXT_ERROR_ACCOUNT_BANED), ToolsDate.dateToStringFull(Long.parseLong(ex.params[0])))))
                .send(api);
        return request;
    }

    public static <K extends Request.Response> Request<K> executeProgressDialog(Request<K> request, Callback1<K> onComplete) {
        return executeProgressDialog((String) null, request, onComplete);
    }

    public static <K extends Request.Response> Request<K> executeProgressDialog(@StringRes int title, Request<K> request, Callback1<K> onComplete) {
        return executeProgressDialog(ToolsResources.getString(title), request, onComplete);
    }

    public static <K extends Request.Response> Request<K> executeProgressDialog(String title, Request<K> request, Callback1<K> onComplete) {
        Widget dialog = title == null ? ToolsView.showProgressDialog() : ToolsView.showProgressDialog(title);
        return executeProgressDialog(dialog, request, onComplete);
    }

    public static <K extends Request.Response> Request<K> executeProgressDialog(Widget dialog, Request<K> request, Callback1<K> onComplete) {
        request.onComplete(r -> onComplete.callback(r))
                .onNetworkError(() -> ToolsToast.show(SupAndroid.TEXT_ERROR_NETWORK))
                .onFinish(() -> dialog.hide())
                .onApiError(ApiClient.ERROR_ACCOUNT_IS_BANED, ex -> ToolsToast.show(String.format(ToolsResources.getString(SupAndroid.TEXT_ERROR_ACCOUNT_BANED), ToolsDate.dateToStringFull(Long.parseLong(ex.params[0])))))
                .send(api);
        return request;
    }


    public static <K extends Request.Response> Request<K> executeProgressDialog(Request<K> request, Callback2<Widget, K> onComplete) {
        return executeProgressDialog(null, request, onComplete);
    }

    public static <K extends Request.Response> Request<K> executeProgressDialog(@StringRes int title, Request<K> request, Callback2<Widget, K> onComplete) {
        return executeProgressDialog(ToolsResources.getString(title), request, onComplete);
    }

    public static <K extends Request.Response> Request<K> executeProgressDialog(String title, Request<K> request, Callback2<Widget, K> onComplete) {
        Widget w = title == null ? ToolsView.showProgressDialog() : ToolsView.showProgressDialog(title);
        request.onComplete(r -> onComplete.callback(w, r))
                .onNetworkError(() -> {
                    ToolsToast.show(SupAndroid.TEXT_ERROR_NETWORK);
                    w.hide();
                })
                .onApiError(ApiClient.ERROR_ACCOUNT_IS_BANED, ex -> ToolsToast.show(String.format(ToolsResources.getString(SupAndroid.TEXT_ERROR_ACCOUNT_BANED), ToolsDate.dateToStringFull(Long.parseLong(ex.params[0])))))
                .send(api);
        return request;
    }

    public static <K extends Request.Response> Request<K> executeEnabledCallback(Request<K> request, Callback1<K> onComplete, Callback1<Boolean> enabled) {
        enabled.callback(false);
        request
                .onComplete(r -> onComplete.callback(r))
                .onNetworkError(() -> ToolsToast.show(SupAndroid.TEXT_ERROR_NETWORK))
                .onFinish(() -> enabled.callback(true))
                .onApiError(ApiClient.ERROR_ACCOUNT_IS_BANED, ex -> ToolsToast.show(String.format(ToolsResources.getString(SupAndroid.TEXT_ERROR_ACCOUNT_BANED), ToolsDate.dateToStringFull(Long.parseLong(ex.params[0])))))
                .send(api);

        return request;
    }


    public static <K extends Request.Response> Request<K> executeEnabled(Widget widget, Request<K> request, Callback1<K> onComplete) {
        if (widget != null) widget.setEnabled(false);
        request
                .onComplete(r -> {
                    onComplete.callback(r);
                    if (widget != null) widget.hide();
                })
                .onNetworkError(() -> ToolsToast.show(SupAndroid.TEXT_ERROR_NETWORK))
                .onFinish(() -> {
                    if (widget != null) widget.setEnabled(true);
                    if (widget != null && widget instanceof WidgetProgressTransparent) widget.hide();
                    if (widget != null && widget instanceof WidgetProgressWithTitle) widget.hide();
                })
                .send(api);

        return request;
    }

    public static <K extends Request.Response> Request<K> executeEnabledConfirm(@StringRes int text, @StringRes int enter, Request<K> request, Callback1<K> onComplete) {
        new WidgetAlert()
                .setText(text)
                .setOnCancel(SupAndroid.TEXT_APP_CANCEL)
                .setAutoHideOnEnter(false)
                .setOnEnter(enter, w -> executeEnabled(w, request, onComplete))
                .asDialogShow();
        return request;
    }
    
}

File: DevSupportAndroid/src/main/java/com/sup/dev/android/libs/api_simple/ApiRequestsSupporter.kt
package com.sup.dev.android.libs.api_simple

import android.support.annotation.StringRes
import com.sup.dev.android.R
import com.sup.dev.android.libs.screens.Screen
import com.sup.dev.android.libs.screens.navigator.NavigationAction
import com.sup.dev.android.libs.screens.navigator.Navigator
import com.sup.dev.android.tools.ToolsResources
import com.sup.dev.android.tools.ToolsToast
import com.sup.dev.android.tools.ToolsView
import com.sup.dev.android.views.screens.SAlert
import com.sup.dev.android.views.screens.SInterstitialProgress
import com.sup.dev.android.views.widgets.Widget
import com.sup.dev.android.views.widgets.WidgetAlert
import com.sup.dev.android.views.widgets.WidgetProgressTransparent
import com.sup.dev.android.views.widgets.WidgetProgressWithTitle
import com.sup.dev.java.libs.api_simple.client.ApiClient
import com.sup.dev.java.libs.api_simple.client.Request
import com.sup.dev.java.libs.debug.log
import com.sup.dev.java.tools.ToolsDate

object ApiRequestsSupporter {


    private var api: ApiClient? = null

    fun init(api: ApiClient) {
        ApiRequestsSupporter.api = api
    }

    fun <K : Request.Response> execute(request: Request<K>, onComplete: (K) -> Unit): Request<K> {
        request.onComplete { r -> onComplete.invoke(r) }
                .onNetworkError { ToolsToast.show(R.string.error_network) }
                .onApiError(ApiClient.ERROR_ACCOUNT_IS_BANED) { ex -> ToolsToast.show(String.format(ToolsResources.s(R.string.error_account_baned), ToolsDate.dateToStringFull(java.lang.Long.parseLong(ex.params!![0])))) }
                .onApiError(ApiClient.ERROR_GONE) { ex -> ToolsToast.show(String.format(ToolsResources.s(R.string.error_gone))) }
                .send(api!!)
        return request
    }

    fun <K : Request.Response> executeInterstitial(action: NavigationAction, request: Request<K>, onComplete: (K) -> Screen): Request<K> {
        val sInterstitialProgress = SInterstitialProgress()
        Navigator.action(action, sInterstitialProgress)

        return execute(request) { r ->
            if (Navigator.getCurrent() === sInterstitialProgress) {
                Navigator.replace(onComplete.invoke(r))
            }
        }
                .onApiError(ApiClient.ERROR_GONE) {
                    if (Navigator.getCurrent() === sInterstitialProgress) SAlert.showGone(Navigator.REPLACE)
                }
                .onNetworkError {
                    if (Navigator.getCurrent() === sInterstitialProgress)
                        SAlert.showNetwork(Navigator.REPLACE) {
                            Navigator.replace(sInterstitialProgress)
                            request.send(api!!)
                        }
                }
    }

    fun <K : Request.Response> executeProgressDialog(request: Request<K>, onComplete: (K) -> Unit): Request<K> {
        return executeProgressDialog(null as String?, request, onComplete)
    }

    fun <K : Request.Response> executeProgressDialog(@StringRes title: Int, request: Request<K>, onComplete: (K) -> Unit): Request<K> {
        return executeProgressDialog(ToolsResources.s(title), request, onComplete)
    }

    fun <K : Request.Response> executeProgressDialog(title: String?, request: Request<K>, onComplete: (K) -> Unit): Request<K> {
        val dialog = if (title == null) ToolsView.showProgressDialog() else ToolsView.showProgressDialog(title)
        return executeProgressDialog(dialog, request, onComplete)
    }

    fun <K : Request.Response> executeProgressDialog(dialog: Widget, request: Request<K>, onComplete: (K) -> Unit): Request<K> {
        return execute(request, onComplete)
                .onFinish { dialog.hide() }
    }

    fun <K : Request.Response> executeProgressDialog(request: Request<K>, onComplete: (Widget, K) -> Unit): Request<K> {
        return executeProgressDialog<K>(null, request, onComplete)
    }

    fun <K : Request.Response> executeProgressDialog(@StringRes title: Int, request: Request<K>, onComplete: (Widget, K) -> Unit): Request<K> {
        return executeProgressDialog(ToolsResources.s(title), request, onComplete)
    }

    fun <K : Request.Response> executeProgressDialog(title: String?, request: Request<K>, onComplete: (Widget, K) -> Unit): Request<K> {
        val w = if (title == null) ToolsView.showProgressDialog() else ToolsView.showProgressDialog(title)
        return execute(request) { r -> onComplete.invoke(w, r) }
                .onNetworkError {
                    ToolsToast.show(R.string.error_network)
                    w.hide()
                }
    }

    fun <K : Request.Response> executeEnabledCallback(request: Request<K>, onComplete: (K) -> Unit, enabled: (Boolean) -> Unit): Request<K> {
        enabled.invoke(false)
        return execute(request, onComplete)
                .onFinish { enabled.invoke(true) }
    }


    fun <K : Request.Response> executeEnabled(widget: Widget?, request: Request<K>, onComplete: (K) -> Unit): Request<K> {
        widget?.setEnabled(false)
        widget?.hideCancel()
        return execute(request) { r ->
            onComplete.invoke(r)
            widget?.hideForce()
        }.onFinish {
            widget?.setEnabled(true)
            if (widget != null && widget is WidgetProgressTransparent) widget.hideForce()
            if (widget != null && widget is WidgetProgressWithTitle) widget.hideForce()
        }
    }

    fun <K : Request.Response> executeEnabledConfirm(@StringRes text: Int, @StringRes enter: Int, request: Request<K>, onComplete: (K) -> Unit): Request<K> {
        return executeEnabledConfirm(ToolsResources.s(text), ToolsResources.s(enter), request, onComplete)
    }

    fun <K : Request.Response> executeEnabledConfirm(text: String, enter: String, request: Request<K>, onComplete: (K) -> Unit): Request<K> {
        WidgetAlert()
                .setText(text)
                .setOnCancel(R.string.app_cancel)
                .setAutoHideOnEnter(false)
                .setOnEnter(enter) { w -> executeEnabled(w, request, onComplete) }
                .asSheetShow()
        return request
    }
}